Convert `active` IMAP connection to a pool

acammack-r7 opened the pull request:
The `active` IMAP connection would become confused when used for
concurrent requests, so it has been converted to a pool to allow for
transactional request handling. All internal usage of raw IMAP
connections has been converted to switchboard:with_imap/2.

jtmoulia commented on src/switchboard.erl:
```diff
@@ -198,19 +198,16 @@ await(Account, Type, Timeout) ->
-return(_Account, _IMAP) ->
-    ok.
+return(Account, IMAP) ->
+    ok = poolboy:checkin({via, gproc, key_for(Account, pool)}, IMAP).
```
nitpick: could we have this function still return `ok`? I don't think anything depends on it, but it does change the behaviour.

jtmoulia commented on src/imap.erl:
```diff
@@ -206,6 +206,9 @@ start(ConnSpec, Opts) ->
 %% @equiv start_link(ConnSpec, [])
 -spec start_link(connspec()) ->
     {ok, pid()} | _.
+%% Allow starting with Poolboy
```
super nitpick: mind ending the comment with a `.`? The rest of the codebase uses that convention because it has an Erlang-y feel.

jtmoulia commented on src/switchboard.erl:
```diff
@@ -198,19 +198,16 @@ await(Account, Type, Timeout) ->
 
 
-%% @todo rethrow errors instead of eating them?
+%% @doc Execute a function with an IMAP connection to the given account
```
`.` at the end of the comment

jtmoulia commented on src/switchboard_accounts.erl:
```diff
@@ -67,16 +67,12 @@ init({ConnSpec, Auth, Mailboxes}) ->
+                        Account,
+                        [{name,
+                          {via, gproc, switchboard:key_for(Account, pool)}},
+                         {worker_module, imap}, {size, 4}, {max_overflow, 5}],
```
What do you think about allowing the pool size to be configurable? 4 IMAP connections means that switchboard will be holding open a minimum of 5 per account, which is pretty heavy.

jtmoulia commented:
Thanks a ton for this PR -- this is great. I left some comments, the most important of which is allowing the pool size to be configurable. Sorry, I came in a bit late on this so let me know if you have the bandwidth to address the issues.

Also, this'll result in a minor version bump to `0.4.0`, but I'll commit that separately once merged.

acammack-r7 commented:
@jtmoulia, addressed your comments. I added a application environment variable to configure the pool size and set the default to a more conservative value.

jtmoulia commented:
Merged! I'll create the new release and, if you don't mind, add you to the contributors list.

Cheers 🍻
